### PR TITLE
Add revocation option to the generated CRED_DEF section

### PR DESCRIPTION
**Pull request by swcurran**
Adds to the Generate and Publish CRED_DEF section what happens when the issuer sets `support_revocation` to `true`. Some reorganizing of the CRED_DEF setup section to cover both the without and with revocation.

I did not change all references of CredDef to CRED_DEF, but will likely do that before suggesting this be merged.

### CODE REVIEW

**Comment by swcurran** [comment]
This PR is now ready for review and needed to unblock the PR #36 .

Please review @matt-raffel-kiva @hkny @victormartinez-work

**Comment by swcurran** [comment]
@matt-raffel-kiva @ArPhil @victormartinez-work -- need another review. I had to merge in the updates from earlier.

File: spec/data_flow_setup.md [hunk 1]
```diff
@@ -108,88 +108,103 @@ following is an example Schema:
...
+additions to the generation process and data structures when support for
+revocations is activated for a given [[ref: CRED_DEF]].
+
+##### Retrieving the SCHEMA Object
+
+Prior to creating a [[ref: CRED_DEF]], the Issuer must get an instance of the
+[[ref: SCHEMA]] upon which the [[ref: CRED_DEF]] will be created, including the
+identifier for the [[ref: SCHEMA]] in the form of a Hyperledger Indy `TXN_ID`.
+If the Issuer is also the [[ref: SCHEMA]] Publisher, they will already have the
+[[ref: SCHEMA]] (with `TXN_ID`). If not the Issuer must request that information
```
Comment (victormartinez-work) [inline spec/data_flow_setup.md:161]:
```suggestion
[[ref: SCHEMA]] (with `TXN_ID`). If not, the Issuer must request that information
```

File: spec/data_flow_setup.md [hunk 2]
```diff
@@ -69,24 +69,24 @@ sequenceDiagram
...
+AnonCreds. The [[ref: SCHEMA]] is defined and published by the [[ref: SCHEMA]] Publisher. Any issuer
+who can reference the [[ref: SCHEMA]] (including the [[ref: SCHEMA]] Publisher) MAY issue
+credentials of that type by creating and publishing a [[ref: CRED_DEF]] based on the
+[[ref: SCHEMA]]. This part of the specification covers the operation to create and
+publish a [[ref: SCHEMA]]. The flow of operations to publish a [[ref: SCHEMA]] is illustrated in
+the `SCHEMA Publisher: Publish SCHEMA` section of the [AnonCreds Setup Data
 Flow](#anoncreds-setup-data-flow) sequence diagram.
 
-The Schema is a JSON structure that can be manually constructed containing the
+The [[ref: SCHEMA]] is a JSON structure that can be manually constructed containing the
```
Comment (victormartinez-work) [inline spec/data_flow_setup.md:84]:
```suggestion
The [[ref: SCHEMA]] is a JSON structure that can be manually constructed, containing the
```

File: spec/data_flow_setup.md [hunk 3]
```diff
@@ -221,30 +236,114 @@ The integers shown with ellipses (e.g. `123...789`) are all very long integers o
...
+* `object type`: The type of object. `3` is used for [[ref: CRED_DEF]]s.
+* `signature_type`: The `signature_type` item from the [[ref: CRED_DEF]].
+* `SCHEMA TXN_ID`: The `ref` item from the [[ref: CRED_DEF]]
+* `tag`: The `tag` item from the [[ref: CRED_DEF]].
+
+##### Generating a CRED_DEF With Revocation Support
+
+The issuer enables the ability to revoke credentials produced from a [[ref: CRED_DEF]] by
+passing to the [[ref: CRED_DEF]] generation process the flag `support_revocation` as
+`true`. When revocation is to be supported, additional data related to
```
Comment (victormartinez-work) [inline spec/data_flow_setup.md:262]: When revocation is supported or  when revocation is enabled ?  I can see in other parts also mentioned "revocation enabled".  Maybe we should be consistent and either use supported or enabled.

**Review by victormartinez-work — APPROVED** [review]

File: spec/data_flow_setup.md [hunk 4]
```diff
@@ -69,24 +69,24 @@ sequenceDiagram
     end
 ```
 
-#### Schema Publisher: Publish Schema Object
+#### SCHEMA Publisher: Publish SCHEMA Object
 
-Each type of AnonCred credential is based on a Schema published to a Verifiable
+Each type of AnonCred credential is based on a [[ref: SCHEMA]] published to a Verifiable
 Data Registry (VDR), an instance of Hyperledger Indy in this version of
```
Comment (ArPhil) [inline spec/data_flow_setup.md:75]: I believe we are missing a definition in the terminology section for "Verifiable Data Registry (VDR)" or is ist something that we assume the reader already knows?

File: spec/data_flow_setup.md [hunk 5]
```diff
@@ -69,24 +69,24 @@ sequenceDiagram
...
-AnonCreds. The Schema is defined and published by the Schema Publisher. Any issuer
-who can reference the Schema (including the Schema Publisher) MAY issue
-credentials of that type by creating and publishing a CredDef based on the
-Schema. This part of the specification covers the operation to create and
-publish a Schema. The flow of operations to publish a Schema is illustrated in
-the `Schema Publisher: Publish Schema` section of the [AnonCreds Setup Data
+AnonCreds. The [[ref: SCHEMA]] is defined and published by the [[ref: SCHEMA]] Publisher. Any issuer
```
Comment (ArPhil) [inline spec/data_flow_setup.md:76]:
```suggestion
AnonCreds. The [[ref: SCHEMA]] is defined and published by the [[ref: SCHEMA publisher]]. Any issuer
```

File: spec/data_flow_setup.md [hunk 6]
```diff
@@ -69,24 +69,24 @@ sequenceDiagram
...
+who can reference the [[ref: SCHEMA]] (including the [[ref: SCHEMA]] Publisher) MAY issue
```
Comment (ArPhil) [inline spec/data_flow_setup.md:77]:
```suggestion
who can reference the [[ref: SCHEMA]] (including the [[ref: SCHEMA publisher]]) MAY issue
```

File: spec/data_flow_setup.md [hunk 7]
```diff
@@ -108,88 +108,103 @@ following is an example Schema:
...
+* A link to the [[ref: SCHEMA]] upon which the [[ref: CRED_DEF]] is based (the credential type).
 * A set of public/private key pairs, one per attribute (claim) in the
   credential. The private keys will later be used to sign the claims when
   credentials to be issued are created.
 * Other information necessary for the cryptographic signing of credentials.
+* Information necessary for the revocation of credentials, if revocation is to
+  be supported by the Issuer for this type of credential.
+
+We'll initially cover the generation and data for a [[ref: CRED_DEF]] created without the
+option of revoking credentials. Following that, ([in this
```
Comment (ArPhil) [inline spec/data_flow_setup.md:150]:
```suggestion
option of revoking credentials. Following that, [in this
section](#generating-a-cred_def-with-revocation-support),
```

File: spec/data_flow_setup.md [hunk 8]
```diff
@@ -108,88 +108,103 @@ following is an example Schema:
...
+section](#generating-a-cred_def-with-revocation-support)), we describe the
```
Comment (ArPhil) [inline spec/data_flow_setup.md:151]:
```suggestion
option of revoking credentials. Following that, [in this
section](#generating-a-cred_def-with-revocation-support), we describe the
```

File: spec/data_flow_setup.md [hunk 9]
```diff
@@ -108,88 +108,103 @@ following is an example Schema:
...
+additions to the generation process and data structures when support for
+revocations is activated for a given [[ref: CRED_DEF]].
+
+##### Retrieving the SCHEMA Object
+
+Prior to creating a [[ref: CRED_DEF]], the Issuer must get an instance of the
+[[ref: SCHEMA]] upon which the [[ref: CRED_DEF]] will be created, including the
+identifier for the [[ref: SCHEMA]] in the form of a Hyperledger Indy `TXN_ID`.
+If the Issuer is also the [[ref: SCHEMA]] Publisher, they will already have the
```
Comment (ArPhil) [inline spec/data_flow_setup.md:160]:
```suggestion
If the Issuer is also the [[ref: SCHEMA publisher]], they will already have the
```

File: spec/data_flow_setup.md [hunk 10]
```diff
@@ -221,30 +236,114 @@ The integers shown with ellipses (e.g. `123...789`) are all very long integers o
 * `primary` is the data used for generating credentials.
 * `n` is the ...
 * `r` is a list of the attributes in the credential, with an associated public key for each.
```
Comment (ArPhil) [inline spec/data_flow_setup.md:238]:
```suggestion
* `r` is a list of attribute names which are going to be used in the credential, each with an associated public key.
```

I`d rather emphasize that those are the "attribute names" in contrast to the attribute (values) itselfes.

File: spec/data_flow_setup.md (hunk 3 above)
Comment (ArPhil) [reply spec/data_flow_setup.md:262]: I agree. Per default revocation is supported by Indy. IMO enabling or disabling revocation purely depends on setting the revocation flag when creating the CRED_DEF

Comment (swcurran) [reply spec/data_flow_setup.md:262]: Updated all places in the section where I had used "supported" when it should have been "enabled".  I also replaced an instance of "activated" with "enabled" to be consistent.

File: spec/data_flow_setup.md (hunk 6 above)
Comment (swcurran) [reply spec/data_flow_setup.md:77]: Resolved -- checked the entire document and updated all instances in prose (e.g. excluding the diagram).

File: spec/data_flow_setup.md (hunk 1 above)
Comment (swcurran) [reply spec/data_flow_setup.md:161]: Fixed.

File: spec/data_flow_setup.md (hunk 10 above)
Comment (swcurran) [reply spec/data_flow_setup.md:238]: Good catch. Fixed.

File: spec/data_flow_setup.md [hunk 11]
```diff
@@ -221,30 +236,114 @@ The integers shown with ellipses (e.g. `123...789`) are all very long integers o
...
-  * `master_secret` is in all CredDefs and is for signing attribute that will be put into the credential from the Holder.
-  * The rest of the attributes in the list are those defined in the Schema.
-  * The attribute names are normalized (lower case, spaces removed) and listed in the CredDef in alphabetical order.
+  * `master_secret` is in all [[ref: CRED_DEF]]s and is for signing attribute that will be put into the credential from the Holder.
```
[the remark this answers is not in the document]
Comment (swcurran) [reply spec/data_flow_setup.md:239]: Looks good. I adjusted the phrasing a little more.

**Comment by swcurran** [comment]
Note: Rather than accepting the suggestions, I applied them in my branch.  I'm not quite sure that if I accept the suggestion via the UI that I will be able to properly DCO the resulting commit, and once DCO is broken, I generally have to cancel and resubmit the PR.

Ready for another review.

**Pull request merged**